File: src/main/java/net/matees/arcade/Minigame.java
package net.matees.arcade;

import me.kodysimpson.simpapi.command.SubCommand;
import me.kodysimpson.simpapi.menu.Menu;
import net.matees.settings.Setting;

import org.bukkit.event.Listener;
import org.bukkit.inventory.ItemStack;

import java.util.List;

public abstract class Minigame {
    private boolean isCurrentMinigame;

    public boolean isCurrentMinigame() {
        return this.isCurrentMinigame;
    }

    public void setIsCurrentMinigame(boolean current) {
        this.isCurrentMinigame = current;
    }

    public abstract String getName();

    public abstract MinigameType getMinigameType();

    public abstract List<Listener> getListeners();

    public abstract List<Class<? extends SubCommand>> getSubCommands();

    public abstract String getCommandName();

    public abstract String getCommandDescription();

    public abstract String getCommandUsage();

    public abstract List<Setting> getSettings();

    public abstract Class<? extends Menu> getSettingsMenu();

    public Setting getSetting(String name) {
        for (Setting setting : this.getSettings()) {
            if (setting.getName().equals(name)) {
                return setting;
            }
        }

        return null;
    }

    public abstract ItemStack getMinigameMenuItem();

    public abstract void doStartMinigame();

    public void startMinigame() {
        this.setIsCurrentMinigame(true);
        this.doStartMinigame();
    }

    public void stopMinigame() {
        this.setIsCurrentMinigame(false);
        this.onStopMinigame();
    }

    public abstract void onStopMinigame();

}

File: src/main/java/net/matees/arcade/Minigame.kt
package net.matees.arcade

import me.kodysimpson.simpapi.command.SubCommand
import me.kodysimpson.simpapi.menu.Menu
import net.matees.settings.AbstractSetting
import org.bukkit.Bukkit
import org.bukkit.event.Listener
import org.bukkit.inventory.ItemStack

abstract class Minigame {
    var isCurrentMinigame: Boolean = false

    abstract val name: String

    abstract val minigameType: MinigameType?

    abstract val listeners: List<Listener>

    abstract val subCommands: List<Class<out SubCommand?>?>?

    abstract val commandName: String?

    abstract val commandDescription: String?

    abstract val commandUsage: String?

    abstract val settings: List<AbstractSetting<*>>

    abstract val settingsMenu: Class<out Menu?>?

    fun getSetting(name: String): AbstractSetting<*>? {
        for (setting in this.settings) {
            if (setting.name == name) {
                return setting
            }
        }

        return null
    }

    abstract val minigameMenuItem: ItemStack

    abstract fun doStartMinigame()

    fun startMinigame() {
        this.isCurrentMinigame = true
        this.doStartMinigame()
    }

    fun stopMinigame() {
        this.isCurrentMinigame = false
        this.onStopMinigame()
        Bukkit.getServer().broadcastMessage("Minigame ended boozo's")
    }

    abstract fun onStopMinigame()
}
